server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -162,7 +179,24 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
 
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            IndexItem indexItem = (IndexItem) item;
             if (shardLocation == null) {
+                if (allTargetColumns != null) {
+                    // Validation is done before creating an index in order to ensure
```

Comment:
We have to do validation of all columns and it means we might do it twice, second time on indexing

 However it's true only for tables with partitioned by columns. I was thinking about some optimization for the check to not run if `table.partitionedBy.empty == false` but I think  `shardLocation == null `check already ensures that we do "double" validation only for partitioned tables.

Reply:
Can't we pass in a `BiConsumer<String, IndexItem>` instead to all components instead? Looks to me that this is possible (but maybe I miss something) and would avoid that all the required arguments needed by `checkContraints` are passed through the whole chain.

Reply:
We cannot have a single `BiConsumer` for all components since `InsertFromValues` has it's own values for targetColumns, cache, tableInfo and contexts.

Thus, created dedicated BiConsuer-s for insert-from-subquery and insert-from-values in https://github.com/crate/crate/pull/14373/commits/b2c2d4478779613feaf53f2bbfb84ddd450f53d4.

Got rid from references and tableInfo, thanks!